File: android/app/src/main/java/dev/rahier/pouleparty/ui/gamecreation/steps/ZonesRecapStep.kt
@file:OptIn(com.mapbox.maps.MapboxExperimental::class)

package dev.rahier.pouleparty.ui.gamecreation.steps

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Shuffle
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.mapbox.geojson.Point
import com.mapbox.maps.CameraOptions
import com.mapbox.maps.extension.compose.MapboxMap
import com.mapbox.maps.extension.compose.animation.viewport.rememberMapViewportState
import com.mapbox.maps.extension.compose.annotation.ViewAnnotation
import com.mapbox.maps.extension.compose.annotation.generated.PolylineAnnotation
import com.mapbox.maps.viewannotation.geometry
import com.mapbox.maps.viewannotation.viewAnnotationOptions
import dev.rahier.pouleparty.model.Game
import dev.rahier.pouleparty.model.GameMod
import dev.rahier.pouleparty.ui.components.circlePolygonPoints
import dev.rahier.pouleparty.ui.components.zonePreviewColor
import dev.rahier.pouleparty.ui.components.zoomForRadius
import dev.rahier.pouleparty.ui.gamelogic.computeDebugShiftedCircles
import dev.rahier.pouleparty.ui.theme.CROrange
import dev.rahier.pouleparty.ui.theme.ZoneGreen
import dev.rahier.pouleparty.ui.theme.gameboyStyle

/**
 * PP-13 phase 1 — read-only map preview of the upcoming game. Shows
 * the initial zone circle plus every future shrunk circle (rainbow
 * palette, numbered 1..N so the chicken reads the trajectory in
 * order) and the user-placed start + final pins inside the disc.
 * Hosts the PP-14 Shuffle button which regenerates
 * `Game.zone.driftSeed` (and re-picks the initial center). Phase 2
 * swaps the client-side `computeZoneRadius` +
 * `computeDebugShiftedCircles` calls for a PP-69 Cloud Function
 * response.
 */
@Composable
fun ZonesRecapStep(
    game: Game,
    onEntered: () -> Unit,
    onShuffle: () -> Unit,
    modifier: Modifier = Modifier,
) {
    LaunchedEffect(Unit) { onEntered() }

    val isStay = game.gameModEnum == GameMod.STAY_IN_THE_ZONE
    val previewCircles = remember(
        game.zone.driftSeed,
        game.zone.radius,
        game.zone.center,
        game.zone.finalCenter,
        game.timing.start,
        game.timing.end,
        game.timing.headStartMinutes,
        game.zone.shrinkIntervalMinutes,
        game.zone.shrinkMetersPerUpdate,
    ) {
        if (isStay) computeDebugShiftedCircles(game) else emptyList()
    }
    val mapViewportState = rememberMapViewportState {
        setCameraOptions {
            center(game.initialLocation)
            zoom(zoomForRadius(game.zone.radius * 1.15, game.initialLocation.latitude()).toDouble())
        }
    }
    // Re-fit the camera whenever the computed disc center or radius
    // shifts (entering the recap and shuffling both move them).
    LaunchedEffect(
        game.initialLocation.latitude(),
        game.initialLocation.longitude(),
        game.zone.radius,
    ) {
        mapViewportState.flyTo(
            CameraOptions.Builder()
                .center(game.initialLocation)
                .zoom(zoomForRadius(game.zone.radius * 1.15, game.initialLocation.latitude()).toDouble())
                .build()
        )
    }

    Box(modifier = modifier.fillMaxSize()) {
        MapboxMap(
            modifier = Modifier.fillMaxSize(),
            mapViewportState = mapViewportState,
        ) {
            // Initial radius — neon orange glow (matches PP-11 / PP-12).
            val initialPoints = circlePolygonPoints(game.initialLocation, game.zone.radius)
            PolylineAnnotation(points = initialPoints + listOf(initialPoints.first())) {
                lineColor = CROrange.copy(alpha = 0.15f)
                lineWidth = 8.0
            }
            PolylineAnnotation(points = initialPoints + listOf(initialPoints.first())) {
                lineColor = CROrange.copy(alpha = 0.35f)
                lineWidth = 4.0
            }
            PolylineAnnotation(points = initialPoints + listOf(initialPoints.first())) {
                lineColor = CROrange.copy(alpha = 0.9f)
                lineWidth = 2.0
            }

            // Future shrink circles (stayInTheZone only).
            previewCircles.forEachIndexed { index, circle ->
                val color = zonePreviewColor(index, previewCircles.size)
                val pts = circlePolygonPoints(circle.center, circle.radius)
                PolylineAnnotation(points = pts + listOf(pts.first())) {
                    lineColor = color.copy(alpha = 0.95f)
                    lineWidth = 2.5
                }
            }

            // Final disc glow.
            if (isStay) {
                game.finalLocation?.let { finalPos ->
                    val finalPts = circlePolygonPoints(finalPos, 50.0)
                    PolylineAnnotation(points = finalPts + listOf(finalPts.first())) {
                        lineColor = ZoneGreen.copy(alpha = 0.15f)
                        lineWidth = 8.0
                    }
                    PolylineAnnotation(points = finalPts + listOf(finalPts.first())) {
                        lineColor = ZoneGreen.copy(alpha = 0.5f)
                        lineWidth = 3.0
                    }
                    PolylineAnnotation(points = finalPts + listOf(finalPts.first())) {
                        lineColor = ZoneGreen.copy(alpha = 0.9f)
                        lineWidth = 1.5
                    }
                }
            }

            // PP-13: numbered badge on the initial disc (1) and on
            // every shrink circle (2..N) so the chicken reads the
            // trajectory in order. North-edge anchor keeps them on
            // a single bearing for easy visual scanning.
            ViewAnnotation(
                options = viewAnnotationOptions {
                    geometry(badgeAnchor(game.initialLocation, game.zone.radius, displayIndex = 1))
                    allowOverlap(true)
                }
            ) {
                ShrinkOrderBadge(index = 1, color = CROrange)
            }
            if (isStay) {
                previewCircles.forEachIndexed { index, circle ->
                    ViewAnnotation(
                        options = viewAnnotationOptions {
                            geometry(badgeAnchor(circle.center, circle.radius, displayIndex = index + 2))
                            allowOverlap(true)
                        }
                    ) {
                        ShrinkOrderBadge(
                            index = index + 2,
                            color = zonePreviewColor(index, previewCircles.size),
                        )
                    }
                }
            }

            // User-placed pins — distinct from the geometric initial
            // center which PP-13 picks elsewhere.
            ViewAnnotation(
                options = viewAnnotationOptions {
                    geometry(game.startPinPoint)
                    allowOverlap(true)
                }
            ) {
                MapLabel(text = "START", background = CROrange, textColor = Color.White)
            }

            if (isStay) {
                game.finalLocation?.let { finalPos ->
                    ViewAnnotation(
                        options = viewAnnotationOptions {
                            geometry(finalPos)
                            allowOverlap(true)
                        }
                    ) {
                        MapLabel(text = "FINAL", background = ZoneGreen, textColor = Color.Black)
                    }
                }
            }
        }

        // Shuffle button — stayInTheZone only.
        if (isStay) {
            Box(
                modifier = Modifier
                    .align(Alignment.BottomCenter)
                    .padding(bottom = 24.dp)
                    .shadow(4.dp, RoundedCornerShape(50.dp))
                    .clip(RoundedCornerShape(50.dp))
                    .background(CROrange)
                    .clickable { onShuffle() }
                    .padding(horizontal = 20.dp, vertical = 10.dp),
            ) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Icon(
                        Icons.Default.Shuffle,
                        contentDescription = null,
                        tint = Color.Black,
                        modifier = Modifier.size(18.dp),
                    )
                    Spacer(Modifier.width(6.dp))
                    Text("Shuffle", style = gameboyStyle(12), color = Color.Black)
                }
            }
        }
    }
}

/**
 * PP-13 numbered badge attached to each shrink circle so the chicken
 * reads the shrink order at a glance. The fill colour matches the
 * circle's stroke palette so users can pair label and outline.
 */
@Composable
private fun ShrinkOrderBadge(index: Int, color: Color) {
    Box(
        modifier = Modifier
            .shadow(2.dp, CircleShape)
            .clip(CircleShape)
            .background(color)
            .border(1.dp, Color.White, CircleShape)
            .size(22.dp),
        contentAlignment = Alignment.Center,
    ) {
        Text("$index", style = gameboyStyle(9), color = Color.Black)
    }
}

@Composable
private fun MapLabel(text: String, background: Color, textColor: Color) {
    Box(
        modifier = Modifier
            .clip(RoundedCornerShape(50.dp))
            .background(background)
            .padding(horizontal = 8.dp, vertical = 4.dp),
    ) {
        Text(text, style = gameboyStyle(7), color = textColor)
    }
}

/**
 * Places the numbered badge on the circle's outline at a stable
 * pseudo-random angle in the **NW quadrant** (270° → 360°, i.e.
 * west through north). Same circle index → same bearing on every
 * render — Shuffle doesn't move the badges, only the circles
 * themselves.
 */
private fun badgeAnchor(center: Point, radius: Double, displayIndex: Int): Point {
    val bearingDeg = badgeBearingDegrees(displayIndex)
    val bearingRad = bearingDeg * Math.PI / 180.0
    val metersPerDegLat = 111_111.0
    val cosLat = kotlin.math.cos(center.latitude() * Math.PI / 180.0)
    val metersPerDegLng = if (cosLat == 0.0) metersPerDegLat else 111_111.0 * cosLat
    val dLat = (radius * kotlin.math.cos(bearingRad)) / metersPerDegLat
    val dLng = (radius * kotlin.math.sin(bearingRad)) / metersPerDegLng
    return Point.fromLngLat(center.longitude() + dLng, center.latitude() + dLat)
}

/**
 * Pure-function bearing in `[270°, 360°)` (north-west quadrant)
 * derived from `displayIndex` alone via splitmix64.
 */
private fun badgeBearingDegrees(displayIndex: Int): Double {
    var state = displayIndex.toLong() * -7046029254386353131L // 0x9E3779B97F4A7C15
    if (state == 0L) state = 1L
    state += -7046029254386353131L
    var z = state
    z = (z xor (z ushr 30)) * -4658895280553007687L // 0xBF58476D1CE4E5B9
    z = (z xor (z ushr 27)) * -7723592293110705685L // 0x94D049BB133111EB
    z = z xor (z ushr 31)
    val r = (z.toULong().toDouble()) / ULong.MAX_VALUE.toDouble()
    return 270.0 + r * 90.0
}

File: ios/PouleParty/Features/GameCreation/ZonesRecapStep.swift
//
//  ZonesRecapStep.swift
//  PouleParty
//
//  PP-13 phase 1 — third sub-step of the zone setup flow. Renders a
//  read-only map preview of the initial circle + every future shrunk
//  circle (rainbow palette, numbered) so the chicken can sanity-check
//  the trajectory before locking in the timing. PP-14 phase 1 ships
//  the Shuffle button inside the same view (stayInTheZone only).
//
//  Phase 2 (post-PP-69) will swap the client-side `computeZoneRadius`
//  + `computeDebugShiftedCircles` calls for a CF response.
//

import ComposableArchitecture
import CoreLocation
import MapboxMaps
import SwiftUI

struct ZonesRecapStep: GameCreationStepView {
    static let step: GameCreationStep = .zonesRecap
    @Bindable var store: StoreOf<GameCreationFeature>

    @State private var viewport: Viewport = .camera(
        center: CLLocationCoordinate2D(
            latitude: AppConstants.defaultLatitude,
            longitude: AppConstants.defaultLongitude
        ),
        zoom: 13
    )

    private var previewCircles: [DebugShrinkCircle] {
        computeDebugShiftedCircles(game: store.currentGame)
    }

    private var initialRadius: Double { store.currentGame.zone.radius }
    private var initialCenter: CLLocationCoordinate2D { store.currentGame.initialLocation }
    private var startPin: CLLocationCoordinate2D { store.currentGame.startPinLocation }
    private var finalCenter: CLLocationCoordinate2D? { store.currentGame.finalLocation }
    private var isStayInTheZone: Bool { store.currentGame.gameMode == .stayInTheZone }

    var body: some View {
        ZStack {
            Map(viewport: $viewport) {
                initialRingContent
                shrinkCirclesContent
                finalGlowContent
                badgeAnnotationsContent
                pinAnnotationsContent
            }
            .mapStyle(.streets)
            .ignoresSafeArea()

            shuffleButton
        }
        .onAppear {
            store.send(.zonesRecapEntered)
            // Defer focus to the next runloop tick so the
            // `zonesRecapEntered` mutation (radius + drift seed +
            // computed center) is propagated into our `initialCenter`
            // / `initialRadius` reads before we recenter — otherwise
            // we'd fit on the stale defaults from the previous step.
            Task { @MainActor in
                focusViewport()
            }
        }
        .onChange(of: initialCenter.latitude) { _, _ in focusViewport() }
        .onChange(of: initialCenter.longitude) { _, _ in focusViewport() }
        .onChange(of: initialRadius) { _, _ in focusViewport() }
    }

    // MARK: - Map content (split to keep each closure small enough for
    // the SwiftUI / MapContent type checker)

    @MapContentBuilder
    private var initialRingContent: some MapContent {
        let polygon = Polygon(center: initialCenter, radius: initialRadius, vertices: 72)
        let coords = polygon.outerRing.coordinates
        PolylineAnnotation(lineCoordinates: coords)
            .lineColor(StyleColor(UIColor(Color.CROrange).withAlphaComponent(0.15)))
            .lineWidth(8)
        PolylineAnnotation(lineCoordinates: coords)
            .lineColor(StyleColor(UIColor(Color.CROrange).withAlphaComponent(0.35)))
            .lineWidth(4)
        PolylineAnnotation(lineCoordinates: coords)
            .lineColor(StyleColor(UIColor(Color.CROrange).withAlphaComponent(0.9)))
            .lineWidth(2)
    }

    @MapContentBuilder
    private var shrinkCirclesContent: some MapContent {
        if isStayInTheZone {
            zonePreviewCirclesContent(circles: previewCircles)
        }
    }

    @MapContentBuilder
    private var finalGlowContent: some MapContent {
        if isStayInTheZone, let finalCenter {
            PouleParty.finalZoneGlowContent(center: finalCenter)
        }
    }

    @MapContentBuilder
    private var badgeAnnotationsContent: some MapContent {
        // Each badge lands on its circle's outline at a stable
        // pseudo-random angle in the NW quadrant (north → west) so
        // the chicken always knows where to look. Angle is a pure
        // function of `displayIndex` — Shuffle doesn't move the
        // badges, only the circles themselves.
        MapViewAnnotation(coordinate: badgeAnchor(
            center: initialCenter,
            radius: initialRadius,
            displayIndex: 1
        )) {
            ShrinkOrderBadge(index: 1, color: Color.CROrange)
        }
        .allowOverlap(true)

        if isStayInTheZone {
            ForEvery(Array(previewCircles.enumerated()), id: \.offset) { pair in
                let badgeCoord = badgeAnchor(
                    center: pair.element.center,
                    radius: pair.element.radius,
                    displayIndex: pair.offset + 2
                )
                MapViewAnnotation(coordinate: badgeCoord) {
                    ShrinkOrderBadge(
                        index: pair.offset + 2,
                        color: zonePreviewColor(
                            forIndex: pair.offset,
                            totalCount: previewCircles.count
                        )
                    )
                }
                .allowOverlap(true)
            }
        }
    }

    /// Places the numbered badge on the circle's outline at a
    /// stable pseudo-random angle in the **NW quadrant**
    /// (270° → 360°, i.e. west through north). Same circle index
    /// → same bearing on every render, so the chicken's eye
    /// learns where each badge sits.
    private func badgeAnchor(
        center: CLLocationCoordinate2D,
        radius: Double,
        displayIndex: Int
    ) -> CLLocationCoordinate2D {
        let bearingDeg = badgeBearingDegrees(displayIndex: displayIndex)
        let bearingRad = bearingDeg * .pi / 180
        let metersPerDegLat = 111_111.0
        let cosLat = cos(center.latitude * .pi / 180)
        let metersPerDegLng = cosLat == 0 ? metersPerDegLat : 111_111 * cosLat
        let dLat = (radius * cos(bearingRad)) / metersPerDegLat
        let dLng = (radius * sin(bearingRad)) / metersPerDegLng
        return CLLocationCoordinate2D(
            latitude: center.latitude + dLat,
            longitude: center.longitude + dLng
        )
    }

    /// Pure-function bearing in `[270°, 360°)` (north-west quadrant)
    /// derived from `displayIndex` alone via splitmix64.
    private func badgeBearingDegrees(displayIndex: Int) -> Double {
        var state = UInt64(displayIndex) &* 0x9E3779B97F4A7C15
        if state == 0 { state = 1 }
        state &+= 0x9E3779B97F4A7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58476D1CE4E5B9
        z = (z ^ (z >> 27)) &* 0x94D049BB133111EB
        z = z ^ (z >> 31)
        let r = Double(z) / Double(UInt64.max)
        return 270 + r * 90
    }

    @MapContentBuilder
    private var pinAnnotationsContent: some MapContent {
        MapViewAnnotation(coordinate: startPin) {
            ZoneRecapPinLabel(text: "START", background: Color.CROrange, textColor: .white)
        }
        .allowOverlap(true)

        if isStayInTheZone, let finalCenter {
            MapViewAnnotation(coordinate: finalCenter) {
                ZoneRecapPinLabel(text: "FINAL", background: Color.zoneGreen, textColor: .black)
            }
            .allowOverlap(true)
        }
    }

    // MARK: - Chrome overlays

    @ViewBuilder
    private var shuffleButton: some View {
        if isStayInTheZone {
            VStack {
                Spacer()
                Button {
                    store.send(.shuffleDriftSeed)
                } label: {
                    HStack(spacing: 6) {
                        Image(systemName: "shuffle")
                        Text("Shuffle")
                            .font(.gameboy(size: 12))
                    }
                    .foregroundStyle(.black)
                    .padding(.horizontal, 20)
                    .padding(.vertical, 10)
                    .background(Capsule().fill(Color.CROrange))
                    .shadow(color: .black.opacity(0.2), radius: 4, y: 2)
                }
                .padding(.bottom, 24)
            }
        }
    }

    private func focusViewport() {
        // Mapbox's `withViewportAnimation` sometimes swallows the
        // update when called from `onAppear` before the binding has
        // committed — set the viewport directly first, then animate.
        let targetZoom = zoomForRadius(initialRadius * 1.15, latitude: initialCenter.latitude)
        let target = Viewport.camera(center: initialCenter, zoom: targetZoom)
        viewport = target
        withViewportAnimation(.default(maxDuration: 0.6)) {
            viewport = target
        }
    }
}

/// PP-13 numbered badge attached to each shrink circle so the chicken
/// reads the shrink order at a glance. The fill colour matches the
/// circle's stroke palette so users can pair label and outline.
private struct ShrinkOrderBadge: View {
    let index: Int
    let color: Color

    var body: some View {
        Text("\(index)")
            .font(.gameboy(size: 9))
            .foregroundStyle(.black)
            .frame(minWidth: 18, minHeight: 18)
            .padding(.horizontal, 4)
            .background(
                Capsule()
                    .fill(color)
                    .shadow(color: .black.opacity(0.25), radius: 2, y: 1)
            )
            .overlay(Capsule().stroke(.white, lineWidth: 1))
    }
}

private struct ZoneRecapPinLabel: View {
    let text: String
    let background: Color
    let textColor: Color

    var body: some View {
        Text(text)
            .font(.gameboy(size: 7))
            .foregroundStyle(textColor)
            .padding(.horizontal, 8)
            .padding(.vertical, 4)
            .background(Capsule().fill(background))
    }
}
